### DISCUSSION TITLE
Does the EXPORT option allow other values to be exported other than the generic export datasets mentioned in the documentation ?

### DISCUSSION THREAD

**Discussion by keerthankrishna** [react-querybuilder/react-querybuilder#435] [Q&A] [answered]
I'm looking at getting more than what is available during a **JSON** export using query builder. Does Query Builder allow this ?
When using ` formatQuery(query, 'json'); ` I want to get the _id_ value too (_and maybe other input values_) which I pass to the ` <select> ` component for which I use a custom ` FieldSelectorProps `

Here's a **CodeSandBox** link: https://codesandbox.io/s/hardcore-fermat-9pv5yw?file=/src/App.tsx
In my **Operator JSON** as well as the **Field JSON**, I have an **ID** which I want in the JSON export but I'm unable to get it.
![image](https://user-images.githubusercontent.com/24373382/210102519-54c43fc2-d0d5-48ef-9a49-126d28316654.png)

![image](https://user-images.githubusercontent.com/24373382/210102535-74cde1d8-900d-489c-ab6c-3386bd760b79.png)

_Only thing is: For FIELD, I get a dynamically generated ID which is sent by the QueryBuilder library._
![image](https://user-images.githubusercontent.com/24373382/210102184-cc10b257-ccb5-4303-b840-7a9f566976b1.png)

_Reason for this question is this: If I am getting the JSON data from a database and I want to identify the selected fields and operators, it will be helpful for me to export that ID value once it is selected back to the database._

**Comment by jakeboone02** [comment]
I'm not exactly sure what your desired format is. What do you want the output of `formatQuery` to look like for a given query?

**Comment by keerthankrishna** [comment, marked as answer]
I’m wanting the field’s and Operator’s ID that is passed in the input JSON. Taking my example, I’m looking to get _Last Name’s_ ID of **2** and = _Operator’s_ ID of **101** in the ` formatQuery ` JSON export

**Reply by jakeboone02** [reply]
I understand that, but how do you want it represented in the JSON? I'm looking for an example output. As in, show me the actual JSON you want to see as a result of `formatQuery`, not a description of it.

**Reply by jakeboone02** [reply]
You might find this page helpful: https://react-querybuilder.js.org/docs/tips/adding-removing-query-properties

**Comment by keerthankrishna** [comment]
Sorry for the delay - I will take a look at the page you provided..

Here's what I was expecting and I will see if I can get it with the details mentioned in your page.
![image](https://user-images.githubusercontent.com/24373382/210433924-2f26326a-cf8f-40cb-a44e-06a75bf27543.png)

**Reply by jakeboone02** [reply]
Yeah, I think what you're looking for is [`transformQuery`](https://react-querybuilder.js.org/docs/api/misc#transformquery).

CodeSandbox example here: https://codesandbox.io/s/exciting-violet-ot444i?file=/src/App.tsx

**Reply by keerthankrishna** [reply]
What if I don't have a set constant called "_Operators_" but rather use ` getOperators ` ?
I'm able to handle the _field_id_ the same way you did but I'm not sure how to get the _op_id_ as it is being retrieved via ` switch `case with ` getOperators `

See the fork of your CodeSandbox example here: https://codesandbox.io/s/bitter-rgb-ll8l3g-adding-id-to-fields-ll8l3g?file=/src/App.tsx
How do i get the operator ID in this case ?

**Reply by jakeboone02** [reply]
You can call `getOperators` yourself, passing the rule's `field` just like the main component does:

```ts
const transformedQuery = transformQuery(query, {
  ruleProcessor: r => ({
    ...r,
    field_id: fields.find(f => f.name === r.field)!.id,
    op_id: getOperators(r.field).find(op => op.name === r.operator)!.operator_ID,
  }),
});
```

Also, `getOperators` can be defined outside the component function so it doesn't get recreated on every render.

Corrected CS fork here: https://codesandbox.io/s/qgwtox?file=/src/App.tsx

**Reply by keerthankrishna** [reply]
Thank you so much for this!
` getOperators ` can be defined outside the component function only if Fields data is already set and ready, right ? If not, it has to be defined within the function.

Any idea why field ID and operator ID (_the newly added ID fields in the JSON_) may not be refreshing ? The first selection alone shows me the field and operator IDs and then any additional change or rule additions keeps showing the same field and operator ID values.

**Reply by jakeboone02** [reply]
If `getOperators` needs access to anything scoped within the component function (e.g. your `fields` prop is managed with `useState`), then yes. Otherwise, no.

I'm not sure why you're not seeing the ID's update. They update for me when I test the [sandbox](https://codesandbox.io/s/qgwtox?file=/src/App.tsx).

**Reply by keerthankrishna** [reply]
This is working fine now - not sure what I missed last time.. thank you for checking!
